### ISSUE TITLE
cannot get ip addresses for instances with count/splat

### ISSUE THREAD

**Issue by tomzepf-oracle** [oracle/terraform-provider-baremetal#76]
Instance information returned by the data.baremetal_core_instance data source does not include the IP addresses (or vnic_ids) for the host. This may be due to the underlying API not returning this information directly. Given the limitations of HCL, this appears to makes it impossible to get the IP address(es) for an instance created with a splat/count. Which then appears to make it impossible to use provisioners, etc.

Say a set of bastion instances are created like:

```
/*
 * Bastion instances
 */
resource "baremetal_core_instance" "bastion0" {
  display_name        = "${var.vcn-name}-bastion0"
  hostname_label      = "bastion0"
  compartment_id      = "${var.compartment_ocid}"
  availability_domain = "${element(var.subnet-ads, count.index)}"
  subnet_id           = "${element(var.subnet-ids, count.index)}"
  image               = "${lookup(data.baremetal_core_images.os-images.images[0], "id")}"
  shape               = "${var.compute-shape}"

  metadata {
    ssh_authorized_keys = "${file(var.public_key_path)}"
    user_data           = "${base64encode(file("${path.module}/userdata.sh"))}"
  }
}

resource "baremetal_core_instance" "bastion" {
  display_name        = "${var.vcn-name}-bastion${format("%d", count.index + 1)}"         # move past bastion0
  hostname_label      = "bastion${format("%d", count.index + 1)}"                         # move past bastion0
  compartment_id      = "${var.compartment_ocid}"
  availability_domain = "${element(var.subnet-ads, count.index + 1)}"                     # move past bastion0
  subnet_id           = "${element(var.subnet-ids, count.index + 1)}"                     # move past bastion0
  image               = "${lookup(data.baremetal_core_images.os-images.images[0], "id")}"
  shape               = "${var.compute-shape}"

  count = "${var.bastion-count}"

  metadata {
    ssh_authorized_keys = "${file(var.public_key_path)}"
    user_data           = "${base64encode(file("${path.module}/userdata.sh"))}"
  }
}

/*
 * Bastion IP addresses
 */
data "baremetal_core_vnic_attachments" "bastion-0" {
  compartment_id = "${baremetal_core_instance.bastion0.compartment_id}"
  instance_id    = "${baremetal_core_instance.bastion0.id}"
}

data "baremetal_core_vnic" "bastion-0" {
  vnic_id = "${lookup(data.baremetal_core_vnic_attachments.bastion-0.vnic_attachments[0], "vnic_id")}"
}

data "baremetal_core_vnic_attachments" "bastion" {
  compartment_id = "${element(baremetal_core_instance.bastion.*.compartment_id, count.index)}"
  instance_id    = "${baremetal_core_instance.bastion.id}"
  count          = "${var.bastion-count}"
}

data "baremetal_core_vnic" "bastion" {
  vnic_id = "${lookup(element(data.baremetal_core_vnic_attachments.bastion.*.vnic_attachments[0], count.index), "vnic_id")}"
  # vnic_id = "${lookup(data.baremetal_core_vnic_attachments.bastion.0.vnic_attachments[0], "vnic_id")}"
  count = "${var.bastion-count}"
}

output "bastion-ips" {
  value = [
    "${data.baremetal_core_vnic.bastion-0.public_ip_address}",
    "${data.baremetal_core_vnic.bastion.*.public_ip_address}",
  ]
}

/*
 * Bastion OS image
 */
data "baremetal_core_images" "os-images" {
  compartment_id           = "${var.compartment_ocid}"
  operating_system         = "${var.os-name}"
  operating_system_version = "${var.os-version}"
}
```

The data.baremetal_core_vnic.bastion count/splat data source fails with the message:

```
Error refreshing state: 1 error(s) occurred:

* At column 3, line 1: lookup: argument 1 should be type map, got type string in:

${lookup(element(data.baremetal_core_vnic_attachments.bastion.*.vnic_attachments[0], count.index), "vnic_id")}
```

Perhaps there's some other way to do this, but it does not appear obvious.

**Comment by jvcalise** [comment]
This is a known issue today, and I am tracking this with issue - https://github.com/hashicorp/terraform/issues/13504

The workaround is to move your compute instance creation, along with the vnic_attachment, and vnic lookup to a module, and define the module output to be the vnic public/private IP address.  Output variables allow the splat reference to work correctly.    I am going to test with the latest provider and TF version 9 to see if the problem has been resolved shortly

**Comment by tomzepf-oracle** [comment]
I have prototyped the workaround and it does appear to be an approach to get the IP addresses from outside a module. But this is very inconvenient for the purpose of provisioners, etc, that should be tightly bound to a module. To be more usable, I think it would be best to have a list of VNICs directly accessible from the baremetal_core_instance resource, even if this means hiding some of the complexity in the provider.

**Comment by shomron** [comment]
I have run into the same issue - the current approach unnecessarily exposes us to  complexities of the underlying object model, and is limiting in its expressiveness.

I've prototyped a modification to the provider which exposes `public_ip`, `private_ip` of the first VNIC as computed attributes of the`baremetal_core_instance` resource.

It seems promising so far - I need to tidy up and fix the tests, but I can post here afterwards and solicit feedback :)

**Comment by craigcarl-oracle** [comment]
@shomron Please do submit a pull request. We have been having a discussion internally around if/how much TF resources should deviate from the BMC API, it's an open topic. My current thinking is that, if we decided to value add like this, it would be implemented as a different resource and not a change to the existing one.

**Comment by shomron** [comment]
@craigcarl-oracle roger that. I've opened #94 as a concrete point of discussion - have a look and let's take it from there.

**Comment by codycushing** [comment]
#94 integrated, `baremetal_core_instance...public_ip|private_ip` will be available in v1.0.10. Thanks @shomron
